apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts

```diff
@@ -7,14 +8,80 @@ import asyncHandler from 'express-async-handler';
 
 import * as error from '@prairielearn/error';
 import * as sqldb from '@prairielearn/postgres';
+import { run } from '@prairielearn/run';
 
+import type { PageAuthzData } from '../../lib/authz-data-lib.js';
 import { makeS3ClientConfig } from '../../lib/aws.js';
+import type { RawStaffUser } from '../../lib/client/safe-db-types.js';
 
-import { GradingJobRowSchema, InstructorGradingJob } from './instructorGradingJob.html.js';
+import {
+  type GradingJobRow,
+  GradingJobRowSchema,
+  InstructorGradingJob,
+} from './instructorGradingJob.html.js';
 
 const router = Router();
 const sql = sqldb.loadSqlEquiv(import.meta.url);
 
+/**
+ * Asserts that the user has access to the grading job.
+ *
+ * Throws an error if the user does not have access to the grading job.
+ */
+function assertHasAccessToGradingJob(authzData: PageAuthzData, gradingJobRow: GradingJobRow) {
+  const courseAuthorized = authzData.has_course_permission_preview;
+
+  // 'has_course_instance_permission_view' may be null on course pages.
+  const courseInstanceAuthorized = authzData.has_course_instance_permission_view ?? false;
+
+  const needsStudentDataViewerAccess = userNeedsStudentDataViewerAccess(
+    authzData.user,
+    gradingJobRow,
+  );
+
+  const authorized = run(() => {
+    if (needsStudentDataViewerAccess) {
+      return courseAuthorized && courseInstanceAuthorized;
+    }
+    return courseAuthorized || courseInstanceAuthorized;
+  });
+
+  if (!authorized) {
+    if (needsStudentDataViewerAccess) {
+      throw new error.HttpStatusError(
+        403,
+        'Access denied (must be a course previewer and student data viewer)',
+      );
+    }
+    throw new error.HttpStatusError(
+      403,
+      'Access denied (must be a course previewer or student data viewer)',
```

Comment:
To me, it doesn't seem like student data viewer permissions should be sufficient to access the grading job results for any non-instance-question grading jobs.